File: app/src/main/java/tv/superawesome/demoapp/MainActivity.java
package tv.superawesome.demoapp;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Arrays;
import java.util.List;

import tv.superawesome.lib.sabumperpage.SABumperPage;
import tv.superawesome.sdk.publisher.AwesomeAds;
import tv.superawesome.sdk.publisher.SABannerAd;
import tv.superawesome.sdk.publisher.SAEvent;
import tv.superawesome.sdk.publisher.SAInterface;
import tv.superawesome.sdk.publisher.SAInterstitialAd;
import tv.superawesome.sdk.publisher.SAVideoAd;

public class MainActivity extends Activity {

  /** the options list */
  @Override
  protected void onCreate(final Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    SABumperPage.overrideName("Test app");

    final SABannerAd bannerAd = findViewById(R.id.preview);
    bannerAd.setListener(
        (SAInterface) (placementId, event) -> {
          if(event == SAEvent.adLoaded)
              bannerAd.play(this);
          });

    ListView myList = findViewById(R.id.MyList);
    final List<AdapterItem> data =
        Arrays.asList(
            new HeaderItem("Banners"),
            new PlacementItem("Mobile Leaderboard", 82088, Type.BANNER),
            new HeaderItem("Interstitials"),
            new PlacementItem("Mobile Interstitial Portrait", 82089, Type.INTERSTITIAL),
            new PlacementItem("Interstitial via KSF", 82063, Type.INTERSTITIAL),
            new HeaderItem("Videos"),
            new PlacementItem("Direct Video", 82090, Type.VIDEO),
            new PlacementItem("VPAID via KSF", 82064, Type.VIDEO));
    ListAdapter<AdapterItem> adapter = new ListAdapter<>(this);
    myList.setAdapter(adapter);
    adapter.updateData(data);
    adapter.reloadList();

    myList.setOnItemClickListener(
        (parent, view, position, id) -> {
          AdapterItem item = data.get(position);
          if (item instanceof PlacementItem) {
            PlacementItem placement = (PlacementItem) item;

            switch (placement.type) {
              case BANNER:
                bannerAd.load(placement.pid);
                break;
              case INTERSTITIAL:
                SAInterstitialAd.load(placement.pid, MainActivity.this);
                break;

              case VIDEO:
                SAVideoAd.load(placement.pid, this);
            }
          }
        });

    SAInterstitialAd.setListener(
        (SAInterface)
            (placementId, event) -> {
              if (event == SAEvent.adLoaded) {
                SAInterstitialAd.play(placementId, MainActivity.this);
              }
            });

    SAVideoAd.setListener(
        (SAInterface)
            (placementId, event) -> {
              if (event == SAEvent.adLoaded) {
                SAVideoAd.play(placementId, MainActivity.this);
              }
            });
  }

  public void gotoAdMob(View view) {
    Intent intent = new Intent(this, AdMobActivity.class);
    this.startActivity(intent);
  }

  public void gotoMoPub(View view) {
    Intent intent = new Intent(this, NewTestActivity.class);
    this.startActivity(intent);
  }

  public void ageCheck(View view) {

    final String dateOfBirth = "2012-02-02";

    AwesomeAds.triggerAgeCheck(
        this,
        dateOfBirth,
        isMinorModel -> {
          String message;
          if (isMinorModel != null) {
            message =
                "Min age for '"
                    + isMinorModel.getCountry()
                    + "' is '"
                    + isMinorModel.getConsentAgeForCountry()
                    + "' "
                    + "\nThe age is '"
                    + isMinorModel.getAge()
                    + "' "
                    + "'.\nIs '"
                    + dateOfBirth
                    + "' a minor? -> '"
                    + isMinorModel.isMinor()
                    + "'";
          } else {
            message = "Oops! Something went wrong. No valid model for 'Age Check'...";
          }

          Toast.makeText(MainActivity.this, message, Toast.LENGTH_LONG).show();
        });
  }
}

enum Type {
  BANNER,
  INTERSTITIAL,
  VIDEO
}

class AdapterItem {
  // do nothing
}

class HeaderItem extends AdapterItem {
  public final String title;

  public HeaderItem(String title) {
    this.title = title;
  }
}

class PlacementItem extends AdapterItem {
  public final String name;
  public final int pid;
  public final Type type;

  public PlacementItem(String name, int pid, Type type) {
    this.name = name;
    this.pid = pid;
    this.type = type;
  }
}

class ListAdapter<T extends AdapterItem> extends ArrayAdapter<T> {

  private List<T> data;

  public ListAdapter(Context context) {
    super(context, 0);
  }

  public void updateData(List<T> newData) {
    data = newData;
  }

  public void reloadList() {
    notifyDataSetChanged();
  }

  @Nullable
  @Override
  public T getItem(int i) {
    return data.get(i);
  }

  @Override
  public int getViewTypeCount() {
    return 2;
  }

  @Override
  public int getItemViewType(int i) {
    return getItem(i) instanceof HeaderItem ? 0 : 1;
  }

  @Override
  public int getCount() {
    return data != null ? data.size() : 0;
  }

  @NonNull
  @Override
  public View getView(int position, View convertView, @NonNull ViewGroup parent) {

    if (convertView == null) {
      convertView =
          LayoutInflater.from(getContext()).inflate(R.layout.row_placement, parent, false);
    }

    TextView title = convertView.findViewById(R.id.RowTitle);

    T item = getItem(position);

    if (item instanceof HeaderItem) {
      HeaderItem header = (HeaderItem) item;
      title.setText(header.title);
      convertView.setBackgroundColor(Color.LTGRAY);
    } else {
      PlacementItem placement = (PlacementItem) item;
      if (placement != null) {
        String titleText = placement.pid + " | " + placement.name;
        title.setText(titleText);
      }
      title.setBackgroundColor(Color.WHITE);
    }

    return convertView;
  }
}

File: app/src/main/java/tv/superawesome/demoapp/MainActivity.kt
package tv.superawesome.demoapp

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.fragment.app.FragmentActivity
import com.google.firebase.database.*
import com.google.firebase.ktx.Firebase
import com.google.firebase.database.ktx.database
import kotlinx.android.synthetic.main.activity_main.*
import tv.superawesome.demoapp.adapter.*
import tv.superawesome.demoapp.model.SettingsData
import tv.superawesome.lib.sabumperpage.SABumperPage
import tv.superawesome.lib.sasession.defines.SAConfiguration
import tv.superawesome.sdk.publisher.SAEvent
import tv.superawesome.sdk.publisher.SAInterstitialAd
import tv.superawesome.sdk.publisher.SAVersion
import tv.superawesome.sdk.publisher.SAVideoAd
import tv.superawesome.sdk.publisher.state.CloseButtonState

var data: List<AdapterItem> = listOf()
    private set
private var adapter: CustomListAdapter<AdapterItem>? = null

class MainActivity : FragmentActivity() {

    private lateinit var database: DatabaseReference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        SABumperPage.overrideName("AwesomeAds Demo")

        initUI()

        configureDataSource()
        configureBannerAd()
        configureInterstitialAd()
        configureVideoAd()

        initButtons()
    }

    private fun initButtons() {
        settingsButton.setOnClickListener {
            val dialog = SettingsDialogFragment()
            dialog.show(supportFragmentManager, "settings")
            dialog.onDismissListener = {
                updateSettings()
            }
        }
    }

    private fun getSettings(): SettingsData? = (application as? MyApplication)?.settings

    private fun updateSettings() {
        val settings = getSettings() ?: return
        val config = settings.environment

        bannerView.setConfiguration(config)
        bannerView.setBumperPage(settings.bumperEnabled)
        bannerView.setParentalGate(settings.parentalEnabled)

        SAInterstitialAd.setConfiguration(config)
        SAInterstitialAd.setBumperPage(settings.bumperEnabled)
        SAInterstitialAd.setParentalGate(settings.parentalEnabled)

        SAVideoAd.setConfiguration(config)
        SAVideoAd.setBumperPage(settings.bumperEnabled)
        SAVideoAd.setParentalGate(settings.parentalEnabled)
        SAVideoAd.setMuteOnStart(settings.muteOnStart)

        when (settings.closeButtonState) {
            CloseButtonState.VisibleImmediately -> {
                SAVideoAd.enableCloseButtonNoDelay()
                SAInterstitialAd.enableCloseButtonNoDelay()
            }
            CloseButtonState.VisibleWithDelay -> {
                SAVideoAd.enableCloseButton()
                SAInterstitialAd.enableCloseButton()
            }
            CloseButtonState.Hidden -> SAVideoAd.disableCloseButton()
        }
    }

    private fun initUI() {
        val title = "AwesomeAds: v${SAVersion.getSDKVersionNumber()}"
        titleTextView.text = title
        configureListView()
    }

    private fun configureDataSource() {
        database = Firebase.database(DATABASE_URL).reference

        database.child("list-items").addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                data = dataSnapshot.children.mapNotNull { item ->

                    val rowType = item.getValue(ListItem::class.java)?.rowStyle

                    if(rowType == RowStyle.HEADER) {
                        item.getValue(HeaderItem::class.java)
                    } else {
                        item.getValue(PlacementItem::class.java)
                    }
                }

                adapter?.updateData(data)
                adapter?.reloadList()
            }

            override fun onCancelled(error: DatabaseError) {
                Log.w(TAG, "Failed to load list items.", error.toException())
            }
        })
    }

    private fun configureListView() {
        adapter = CustomListAdapter(this)
        listView.adapter = adapter

        listView.setOnItemClickListener { _, _, position, _ ->
            (data[position] as? PlacementItem)?.let { item ->
                when (item.type) {
                    Type.BANNER -> {
                        if (item.isFull()) {
                            bannerView.load(
                                item.placementId,
                                item.lineItemId ?: 0,
                                item.creativeId ?: 0
                            )
                        } else {
                            bannerView.load(item.placementId)
                        }
                    }
                    Type.INTERSTITIAL -> {
                        if (item.isFull()) {
                            SAInterstitialAd.load(
                                item.placementId,
                                item.lineItemId ?: 0,
                                item.creativeId ?: 0,
                                this@MainActivity
                            )
                        } else {
                            SAInterstitialAd.load(item.placementId, this@MainActivity)
                        }
                    }
                    Type.VIDEO -> {
                        if (item.isFull()) {
                            SAVideoAd.load(
                                item.placementId,
                                item.lineItemId ?: 0,
                                item.creativeId ?: 0,
                                this@MainActivity
                            )
                        } else {
                            SAVideoAd.load(item.placementId, this@MainActivity)
                        }
                    }
                }
            }
        }
    }

    private fun configureVideoAd() {
        SAVideoAd.enableCloseButton()
        SAVideoAd.setListener { placementId, event ->
            updateMessage(placementId, event)
            Log.i(TAG, "SAVideoAd event ${event.name} thread:${Thread.currentThread()}")

            if (event == SAEvent.adLoaded && isPlayEnabled()) {
                SAVideoAd.play(placementId, this@MainActivity)
            }
        }
    }

    private fun configureInterstitialAd() {
        SAInterstitialAd.setListener { placementId, event ->
            updateMessage(placementId, event)
            Log.i(TAG, "SAInterstitialAd event ${event.name} thread:${Thread.currentThread()}")

            if (event == SAEvent.adLoaded && isPlayEnabled()) {
                SAInterstitialAd.play(placementId, this@MainActivity)
            }
        }
    }

    private fun configureBannerAd() {
        bannerView.visibility = View.VISIBLE
        bannerView.setListener { placementId, event ->
            updateMessage(placementId, event)

            if (event == SAEvent.adLoaded && isPlayEnabled()) {
                bannerView.play(this@MainActivity)
            }
        }
    }

    private fun isPlayEnabled(): Boolean {
        return getSettings()?.playEnabled == true
    }

    private fun updateMessage(placementId: Int, event: SAEvent) {
        val settings = getSettings() ?: return

        if(settings.environment == SAConfiguration.UITESTING) {
            val originalMessage = subtitleTextView.text
            val message = "$originalMessage $placementId $event"
            subtitleTextView.text = message
        }
    }
}

private const val TAG = "AwesomeAds"
private const val DATABASE_URL = "https://awesomeads-default-rtdb.europe-west1.firebasedatabase.app/"